src/memo/memo.ml

```diff
@@ -798,21 +799,45 @@ end = struct
       | Needs_work { sample_attempt; _ } -> Running sample_attempt
   end
 
-  let deps_changed =
-    let rec go deps =
-      match deps with
-      | [] -> Changed_or_not.Unchanged
-      | Last_dep.T (dep, v_id) :: deps -> (
-        match Exec_unknown.exec_dep_node_internal_from_sync dep with
-        | Error cycle_error -> Cycle_error cycle_error
-        | Ok res -> (
-          match Value_id.equal res.id v_id with
-          | true -> go deps
-          | false -> Changed ) )
-    in
-    go
-
-  let do_validate (dep_node : _ Dep_node.t) running_state =
+  let restore_from_cache (last_cached_value : _ Cached_value.t option) =
+    match last_cached_value with
+    | None -> Error Cache_lookup_error.Not_found
+    | Some cached_value -> (
+      match cached_value.value with
+      | Cancelled _dependency_cycle ->
+        (* Dependencies of cancelled computations are not accurate, so we can't
+           use [deps_changed] in this case. *)
+        Error Not_found
+      | Error _ ->
+        (* We always recompute errors, so there is no point in checking if any
+           of their dependencies changed. In principle, we could introduce
+           "persistent errors" that are recomputed only when their dependencies
+           have changed. *)
+        Error Not_found
+      | Ok _ -> (
+        let deps_changed =
+          let rec go deps =
+            match deps with
+            | [] -> Changed_or_not.Unchanged
+            | Last_dep.T (dep, v_id) :: deps -> (
+              match Exec_unknown.exec_dep_node_internal_from_sync dep with
+              | Error dependency_cycle -> Cancelled { dependency_cycle }
+              | Ok res -> (
+                match Value_id.equal res.id v_id with
+                | true -> go deps
+                | false -> Changed ) )
+          in
+          go cached_value.deps
+        in
+        match deps_changed with
+        | Unchanged ->
+          cached_value.last_validated_at <- Run.current ();
+          Ok cached_value
+        | Changed -> Error (Out_of_date cached_value)
+        | Cancelled { dependency_cycle } ->
+          Error (Cancelled { dependency_cycle }) ) )
+
+  let compute (dep_node : _ Dep_node.t) cache_lookup_error deps_so_far =
```

Comment:
Something tells me that the term "failure" is better than "error" for things like this where it's completely normal to have failures (also cache lookup failure just sounds familiar somehow).

Reply:
Agreed, I was also thinking about that.